compiler/rustc_expand/src/derive_macro_expansion.rs

```diff
@@ -0,0 +1,125 @@
+use rustc_span::LocalExpnId;
+
+use crate::base::ExtCtxt;
+use crate::errors;
+
+pub(super) fn provide_derive_macro_expansion<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    key: (LocalExpnId, Svh, &'tcx TokenStream),
+) -> Result<&'tcx TokenStream, ()> {
+    let (invoc_id, _macro_crate_hash, input) = key;
+
+    let res = with_context(|(ecx, client)| {
+        let span = invoc_id.expn_data().call_site;
+        let _timer = ecx.sess.prof.generic_activity_with_arg_recorder(
+            "expand_derive_proc_macro_inner",
+            |recorder| {
+                recorder.record_arg_with_span(ecx.sess.source_map(), ecx.expansion_descr(), span);
+            },
+        );
+        let proc_macro_backtrace = ecx.ecfg.proc_macro_backtrace;
+        let strategy = crate::proc_macro::exec_strategy(ecx);
+        let server = crate::proc_macro_server::Rustc::new(ecx);
+        let res = match client.run(&strategy, server, input.clone(), proc_macro_backtrace) {
+            // FIXME(pr-time): without flattened some (weird) tests fail, but no idea if it's correct/enough
+            Ok(stream) => Ok(tcx.arena.alloc(stream.flattened()) as &TokenStream),
+            Err(e) => {
+                ecx.dcx().emit_err({
+                    errors::ProcMacroDerivePanicked {
+                        span,
+                        message: e.as_str().map(|message| errors::ProcMacroDerivePanickedHelp {
+                            message: message.into(),
+                        }),
+                    }
+                });
+                Err(())
+            }
+        };
+        res
+    });
+
+    res
+}
+
+type CLIENT = pm::bridge::client::Client<pm::TokenStream, pm::TokenStream>;
+
+// based on rust/compiler/rustc_middle/src/ty/context/tls.rs
+thread_local! {
+    /// A thread local variable that stores a pointer to the current `CONTEXT`.
+    static TLV: Cell<(*mut (), Option<CLIENT>)> = const { Cell::new((ptr::null_mut(), None)) };
+}
+
+#[inline]
+fn erase(context: &mut ExtCtxt<'_>) -> *mut () {
+    context as *mut _ as *mut ()
+}
+
+#[inline]
+unsafe fn downcast<'a>(context: *mut ()) -> &'a mut ExtCtxt<'a> {
+    unsafe { &mut *(context as *mut ExtCtxt<'a>) }
+}
+
+#[inline]
+fn enter_context_erased<F, R>(erased: (*mut (), Option<CLIENT>), f: F) -> R
+where
+    F: FnOnce() -> R,
+{
+    TLV.with(|tlv| {
+        let old = tlv.replace(erased);
+        let _reset = rustc_data_structures::defer(move || tlv.set(old));
+        f()
+    })
+}
+
+/// Sets `context` as the new current `CONTEXT` for the duration of the function `f`.
+#[inline]
+pub fn enter_context<'a, F, R>(context: (&mut ExtCtxt<'a>, CLIENT), f: F) -> R
+where
+    F: FnOnce() -> R,
+{
+    let (ectx, client) = context;
+    let erased = (erase(ectx), Some(client));
+    enter_context_erased(erased, f)
+}
+
+/// Allows access to the current `CONTEXT` in a closure if one is available.
+#[inline]
+#[track_caller]
+pub fn with_context_opt<F, R>(f: F) -> R
```

Comment:
Could you inline functions that are used once and set visibilities to private when possible?
It will be easier to read that way.

Reply:
Done.